### ISSUE TITLE
Node doesn't add itself to database if a broken node already in database.

### ISSUE THREAD

**Issue by tecnobrat** [Mashape/kong#1160]
We are running 0.8.0 with a postgress backend, running inside docker on elasticbeanstalk.

It looks like if you have a dead node in your database, when you start a new node, it doesn't register itself with postgres, so then you cannot establish a successful cluster.

So we have a down node is in our database (`172.17.55.221`).

```
kong=> select * from nodes;
           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 bfcf7816291c_0.0.0.0:7946 | 172.17.55.221:7946        | 2016-04-18 19:30:27
(1 row)
```

When we start a new node, we get this (I have obfuscated our database info in this log):

```
[ec2-user@ip-172-17-54-190 ~]$ sudo docker logs 847f6171561e
[INFO] kong 0.8.0
[INFO] Using configuration: /etc/kong/kong.yml
[INFO] Setting working directory to /usr/local/kong
[INFO] database...........postgres host=... database=kong user=.. password=... port=5432
[INFO] dnsmasq............address=127.0.0.1:8053 dnsmasq=true port=8053
[INFO] serf ..............-profile=wan -rpc-addr=127.0.0.1:7373 -event-handler=member-join,member-leave,member-failed,member-update,member-reap,user:kong=/usr/local/kong/serf_event.sh -bind=0.0.0.0:7946 -advertise=172.17.54.190:7946 -node=847f6171561e_0.0.0.0:7946 -log-level=err
[INFO] Trying to auto-join Kong nodes, please wait..
[WARN] Cannot join 172.17.55.221:7946. If the node does not exist anymore it will be automatically purged.
[WARN] Could not join the existing cluster
```

And then I check the postgres database again, notice how `172.17.54.190` did not register itself:

```
kong=> select * from nodes;
           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 bfcf7816291c_0.0.0.0:7946 | 172.17.55.221:7946        | 2016-04-18 19:30:27
(1 row)
```

Then I deleted all of the nodes:

```
kong=> delete from nodes;
DELETE 0
kong=> select * from nodes;
 name | cluster_listening_address | created_at
------+---------------------------+------------
```

And restarted kong again:

```
[ec2-user@ip-172-17-54-190 ~]$ sudo docker logs 036e5bdeaa86
[INFO] kong 0.8.0
[INFO] Using configuration: /etc/kong/kong.yml
[INFO] Setting working directory to /usr/local/kong
[INFO] database...........postgres host=... database=kong user=.. password=... port=5432
[INFO] dnsmasq............address=127.0.0.1:8053 dnsmasq=true port=8053
[INFO] serf ..............-profile=wan -rpc-addr=127.0.0.1:7373 -event-handler=member-join,member-leave,member-failed,member-update,member-reap,user:kong=/usr/local/kong/serf_event.sh -bind=0.0.0.0:7946 -advertise=172.17.54.190:7946 -node=036e5bdeaa86_0.0.0.0:7946 -log-level=err
[INFO] Trying to auto-join Kong nodes, please wait..
[WARN] Cannot auto-join the cluster because no nodes were found
```

Now it properly registered itself:

```
kong=> select * from nodes;
           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 036e5bdeaa86_0.0.0.0:7946 | 172.17.54.190:7946        | 2016-04-18 20:42:22
```

And I restarted the other kong node we had:

```
[ec2-user@ip-172-17-50-141 ~]$ sudo docker logs 68f7bf32313a
[INFO] kong 0.8.0
[INFO] Using configuration: /etc/kong/kong.yml
[INFO] Setting working directory to /usr/local/kong
[INFO] database...........postgres host=... database=kong user=.. password=... port=5432
[INFO] dnsmasq............address=127.0.0.1:8053 dnsmasq=true port=8053
[INFO] serf ..............-profile=wan -rpc-addr=127.0.0.1:7373 -event-handler=member-join,member-leave,member-failed,member-update,member-reap,user:kong=/usr/local/kong/serf_event.sh -bind=0.0.0.0:7946 -advertise=172.17.50.141:7946 -node=68f7bf32313a_0.0.0.0:7946 -log-level=err
[INFO] Trying to auto-join Kong nodes, please wait..
[INFO] Successfully auto-joined 172.17.54.190:7946
```

And that one registered as well:

```
kong=> select * from nodes;
           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 68f7bf32313a_0.0.0.0:7946 | 172.17.50.141:7946        | 2016-04-18 20:45:05
 036e5bdeaa86_0.0.0.0:7946 | 172.17.54.190:7946        | 2016-04-18 20:42:22
(2 rows)
```

**Comment by thefosk** [comment]
@tecnobrat I have been trying to replicate this use-case, but I cannot replicate. In all of my tests the new node would always append his advertised address to the database. I have tested this in both Cassandra and PostgreSQL.

#### Kong 1

```yaml
nginx_working_dir: /tmp/kong1
proxy_listen: "0.0.0.0:8000"
proxy_listen_ssl: "0.0.0.0:8443"
admin_api_listen: "0.0.0.0:8001"
cluster_listen: "0.0.0.0:7946"
cluster_listen_rpc: "127.0.0.1:7373"
dns_resolvers_available:
  dnsmasq:
    port: 8053
database: postgres
postgres:
  host: "address"
  port: 5432
  database: "kong"
  user: "kong"
  password: "password"
nginx:
  ...
```

#### Kong 2

```yaml
nginx_working_dir: /tmp/kong2
proxy_listen: "0.0.0.0:9000"
proxy_listen_ssl: "0.0.0.0:9443"
admin_api_listen: "0.0.0.0:9001"
cluster_listen: "0.0.0.0:9946"
cluster_listen_rpc: "127.0.0.1:9373"
dns_resolvers_available:
  dnsmasq:
    port: 9053
database: postgres
postgres:
  host: "address"
  port: 5432
  database: "kong"
  user: "kong"
  password: "password"
nginx:
  ...
```

Then I executed the following operations:

```shell
$ kong start -c kong1.yml # Start first Kong
$ pkill -9 nginx && pkill -9 serf && pkill -9 dnsmasq # Kill the process, so that if results as a failed node
$ kong start -c kong2.yml # Start second Kong
```

Contents of table:

```
kong=> select * from nodes;
            name            | cluster_listening_address |     created_at
----------------------------+---------------------------+---------------------
 local_0.0.0.0:7946 | 192.168.0.156:7946        | 2016-04-20 00:44:03
 local_0.0.0.0:9946 | 192.168.0.156:9946        | 2016-04-20 00:44:22
(2 rows)
```

**Comment by tecnobrat** [comment]
@thefosk I've replicated it again with docker containers.

```
[INFO] kong 0.8.0
[INFO] Using configuration: /etc/kong/kong.yml
[INFO] Setting working directory to /usr/local/kong
[INFO] database...........postgres host=db database=postgres user=postgres port=5432
[INFO] dnsmasq............address=127.0.0.1:8053 dnsmasq=true port=8053
[INFO] serf ..............-profile=wan -rpc-addr=127.0.0.1:7373 -event-handler=member-join,member-leave,member-failed,member-update,member-reap,user:kong=/usr/local/kong/serf_event.sh -bind=0.0.0.0:7946 -node=cd1f8f1604ad_0.0.0.0:7946 -log-level=err
[INFO] Trying to auto-join Kong nodes, please wait..
[WARN] Cannot join 172.17.0.19:7946. If the node does not exist anymore it will be automatically purged.
[WARN] Could not join the existing cluster
[INFO] Auto-generating the default SSL certificate and key...
[INFO] nginx .............admin_api_listen=0.0.0.0:8001 proxy_listen=0.0.0.0:8000 proxy_listen_ssl=0.0.0.0:8443
[OK] Started
```

However, it appears that the new node removes itself very quickly from the cluster, unsure why.... these commands are run within about 2 seconds from eachother, right after I started the new node.

```
postgres=# select * from nodes;
           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 00d912796976_0.0.0.0:7946 | 172.17.0.19:7946          | 2016-04-20 04:06:16
 cd1f8f1604ad_0.0.0.0:7946 | 172.17.0.4:7946           | 2016-04-20 04:07:12
(2 rows)
```
```
postgres=# select * from nodes;
           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 00d912796976_0.0.0.0:7946 | 172.17.0.19:7946          | 2016-04-20 04:06:16
(1 row)
```

**Comment by thefosk** [comment]
@tecnobrat can you send me the exact commands you are running in order to replicate it accurately?

**Comment by tecnobrat** [comment]
I've created a repo which should help you reproduce this bug.

https://github.com/tecnobrat/kong-serf-bug

Here is the output from that script:

```
Killing all potentially conflicting containers
Containers killed
b4cecbfedb430c14737213d05eebf120c63ab3805c9f652a899f63891c152417
Waiting for postgres to start
ERROR:  relation "nodes" does not exist
LINE 1: SELECT * from nodes
                      ^
f4b244d6e7651b22091c62f6179c1f73b1e6057344092e357e13ced44f55903c
Waiting for first kong to boot
==> Starting Serf agent...
==> Starting Serf agent RPC...
==> Serf agent running!
         Node name: 'f4b244d6e765_0.0.0.0:7946'
         Bind addr: '0.0.0.0:7946'
          RPC addr: '127.0.0.1:7373'
         Encrypted: false
          Snapshot: false
           Profile: wan

==> Log data will now stream in as it occurs:

[INFO] kong 0.8.0
[INFO] Using configuration: /etc/kong/kong.yml
[INFO] Setting working directory to /usr/local/kong
[INFO] database...........postgres host=db database=postgres user=postgres password=postgres port=5432
[INFO] Migrating core (postgres)
[INFO] core migrated up to: 2015-01-12-175310_skeleton
[INFO] core migrated up to: 2015-01-12-175310_init_schema
[INFO] core migrated up to: 2015-11-23-817313_nodes
[INFO] core migrated up to: 2016-02-29-142793_ttls
[INFO] Migrating hmac-auth (postgres)
[INFO] hmac-auth migrated up to: 2015-09-16-132400_init_hmacauth
[INFO] Migrating rate-limiting (postgres)
[INFO] rate-limiting migrated up to: 2015-08-03-132400_init_ratelimiting
[INFO] Migrating acl (postgres)
[INFO] acl migrated up to: 2015-08-25-841841_init_acl
[INFO] Migrating oauth2 (postgres)
[INFO] oauth2 migrated up to: 2015-08-03-132400_init_oauth2
[INFO] Migrating jwt (postgres)
[INFO] jwt migrated up to: 2015-06-09-jwt-auth
[INFO] jwt migrated up to: 2016-03-07-jwt-alg
[INFO] Migrating basic-auth (postgres)
[INFO] basic-auth migrated up to: 2015-08-03-132400_init_basicauth
[INFO] Migrating response-ratelimiting (postgres)
[INFO] response-ratelimiting migrated up to: 2015-08-03-132400_init_response_ratelimiting
[INFO] Migrating key-auth (postgres)
[INFO] key-auth migrated up to: 2015-07-31-172400_init_keyauth
[INFO] dnsmasq............address=127.0.0.1:8053 dnsmasq=true port=8053
[INFO] serf ..............-profile=wan -rpc-addr=127.0.0.1:7373 -event-handler=member-join,member-leave,member-failed,member-update,member-reap,user:kong=/usr/local/kong/serf_event.sh -bind=0.0.0.0:7946 -node=f4b244d6e765_0.0.0.0:7946 -log-level=err
[INFO] Trying to auto-join Kong nodes, please wait..
[WARN] Cannot auto-join the cluster because no nodes were found
[INFO] Auto-generating the default SSL certificate and key...
[INFO] nginx .............admin_api_listen=0.0.0.0:8001 proxy_listen=0.0.0.0:8000 proxy_listen_ssl=0.0.0.0:8443
[OK] Started
==> Serf agent running!
         Node name: 'f4b244d6e765_0.0.0.0:7946'
         Bind addr: '0.0.0.0:7946'
          RPC addr: '127.0.0.1:7373'
         Encrypted: false
          Snapshot: false
           Profile: wan

==> Log data will now stream in as it occurs:

Postgres status after first node alive:
           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 f4b244d6e765_0.0.0.0:7946 | 172.17.0.3:7946           | 2016-04-20 22:24:32
(1 row)

Killing node
Waiting incase the node removes itself already
Is bad node still in postgres?
           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 f4b244d6e765_0.0.0.0:7946 | 172.17.0.3:7946           | 2016-04-20 22:24:32
(1 row)

Starting a new node
3f153a9a22dd10516ff949e30a29b35b611ec88604c9825af1344dbe97f8907e
Waiting for new node to boot
[INFO] kong 0.8.0
[INFO] Using configuration: /etc/kong/kong.yml
[INFO] Setting working directory to /usr/local/kong
[INFO] database...........postgres host=db database=postgres user=postgres password=postgres port=5432
[INFO] Migrating core (postgres)
[INFO] core migrated up to: 2015-01-12-175310_skeleton
[INFO] core migrated up to: 2015-01-12-175310_init_schema
[INFO] core migrated up to: 2015-11-23-817313_nodes
[INFO] core migrated up to: 2016-02-29-142793_ttls
[INFO] Migrating hmac-auth (postgres)
[INFO] hmac-auth migrated up to: 2015-09-16-132400_init_hmacauth
[INFO] Migrating rate-limiting (postgres)
[INFO] rate-limiting migrated up to: 2015-08-03-132400_init_ratelimiting
[INFO] Migrating acl (postgres)
[INFO] acl migrated up to: 2015-08-25-841841_init_acl
[INFO] Migrating oauth2 (postgres)
[INFO] oauth2 migrated up to: 2015-08-03-132400_init_oauth2
[INFO] Migrating jwt (postgres)
[INFO] jwt migrated up to: 2015-06-09-jwt-auth
[INFO] jwt migrated up to: 2016-03-07-jwt-alg
[INFO] Migrating basic-auth (postgres)
[INFO] basic-auth migrated up to: 2015-08-03-132400_init_basicauth
[INFO] Migrating response-ratelimiting (postgres)
[INFO] response-ratelimiting migrated up to: 2015-08-03-132400_init_response_ratelimiting
[INFO] Migrating key-auth (postgres)
[INFO] key-auth migrated up to: 2015-07-31-172400_init_keyauth
[INFO] dnsmasq............address=127.0.0.1:8053 dnsmasq=true port=8053
[INFO] serf ..............-profile=wan -rpc-addr=127.0.0.1:7373 -event-handler=member-join,member-leave,member-failed,member-update,member-reap,user:kong=/usr/local/kong/serf_event.sh -bind=0.0.0.0:7946 -node=f4b244d6e765_0.0.0.0:7946 -log-level=err
[INFO] Trying to auto-join Kong nodes, please wait..
[WARN] Cannot auto-join the cluster because no nodes were found
[INFO] Auto-generating the default SSL certificate and key...
[INFO] nginx .............admin_api_listen=0.0.0.0:8001 proxy_listen=0.0.0.0:8000 proxy_listen_ssl=0.0.0.0:8443
[OK] Started
==> Serf agent running!
         Node name: 'f4b244d6e765_0.0.0.0:7946'
         Bind addr: '0.0.0.0:7946'
          RPC addr: '127.0.0.1:7373'
         Encrypted: false
          Snapshot: false
           Profile: wan

==> Log data will now stream in as it occurs:

Checking Postgres
           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 f4b244d6e765_0.0.0.0:7946 | 172.17.0.3:7946           | 2016-04-20 22:24:32
 3f153a9a22dd_0.0.0.0:7946 | 172.17.0.4:7946           | 2016-04-20 22:24:48
(2 rows)

           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 f4b244d6e765_0.0.0.0:7946 | 172.17.0.3:7946           | 2016-04-20 22:24:32
(1 row)

           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 f4b244d6e765_0.0.0.0:7946 | 172.17.0.3:7946           | 2016-04-20 22:24:32
(1 row)

           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 f4b244d6e765_0.0.0.0:7946 | 172.17.0.3:7946           | 2016-04-20 22:24:32
(1 row)

           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 f4b244d6e765_0.0.0.0:7946 | 172.17.0.3:7946           | 2016-04-20 22:24:32
(1 row)

           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 f4b244d6e765_0.0.0.0:7946 | 172.17.0.3:7946           | 2016-04-20 22:24:32
(1 row)

           name            | cluster_listening_address |     created_at
---------------------------+---------------------------+---------------------
 f4b244d6e765_0.0.0.0:7946 | 172.17.0.3:7946           | 2016-04-20 22:24:32
(1 row)

[INFO] Using configuration: /etc/kong/kong.yml
[ERR] serf is not running
[INFO] Using configuration: /etc/kong/kong.yml
3f153a9a22dd_0.0.0.0:7946  172.17.0.4:7946  alive
```

**Comment by asknaskan** [comment]
Having exactly the same problem as ecnobrat described, using kong-docker on ECS Cluster with postgres.

Settings for cluster: 
- `cluster.advertise: "<privateIp>:7946"`
- `cluster_listen: "0.0.0.0:7946"`
- `cluster_listen_rpc: "127.0.0.1:7373"`

**Comment by asknaskan** [comment]
I debugged the code and found out when `/kong/cli/services/serf.lua`->`_add_node()` will be called, that `cluster_utils.get_node_name(self._configuration)` try to get the hostname from the docker and crashes somehow. Investigation goes on.

**Comment by thefosk** [comment]
Do you have any error log?

**Comment by asknaskan** [comment]
Okay, well. Locally I can't reproduce the problem with the docker hostname. It just happens on the EC2 instances. I try to reproduce it there later. For fixing that on the EC2 i've used this in `/tools/utils.lua`

```
function _M.get_hostname()
    local C = ffi.C

    ffi.cdef[[
    int gethostname(char *name, size_t len);
    ]]

    local size = 50
    local buf = ffi.new("unsigned char[?]", size)
    local res = C.gethostname(buf, size)

    if res == 0 then
        local hostname = ffi.string(buf, size)
        hostname = string.gsub(hostname, "%z+$", "")
        return hostname
    end

    local f = io.popen("/bin/hostname", "r")
    if f then
        local hostname = f:read("*a")
        f:close()
        hostname = string.gsub(hostname, "\n$", "")
        return hostname
    end

    return ""
end
```

But anyway this won't help. The code for insert the node to table will be run without any error and the function returns true. The node will not be added to the nodes table in postgres.

**Comment by tecnobrat** [comment]
`/bin/hostname` exists in the docker image, and always returns the container ID, thats normal behaviour for docker containers.

**Comment by dag24** [comment]
I have also seen this problem, but not in Docker. I have no problem with hostnames.

To reproduce, using kong 0.8 and a Postgres 9.5 db:
Have 2 or more kong nodes running in a cluster with each other;
Kill those nodes or `kong stop` them;
Try to `kong start` two fresh nodes (different IPs to the first two);
They will fail to join a cluster with each other.

I work around this by doing `delete from nodes;` in psql and then `kong restart` on the new nodes. They then find each other just fine.

Of note: this problem (new nodes failing to join a cluster) occurs when there exist nodes in the database with the status of `dead` or `left`.

I have a real-life scenario where this is a problem: AWS EC2 auto-scaling group running Kong. Nodes can appear and disappear as part of normal operation.

**Comment by thefosk** [comment]
Could you try to manually apply the patch at #1281 and see if that fixes your problem?

**Comment by thefosk** [comment]
This *should* be fixed in 0.8.3

**Comment by dag24** [comment]
I am still able to reproduce this bug in kong 0.8.3 with Postgres, here are detailed repro steps.

If you do `kong start` and there exists an entry in the `nodes` table with a corresponding entry in `ttls` with an `expire_at` timestamp in the future:
* The node you are starting fails to join the existing cluster (it won't join any *alive* nodes that may also happen to be in the nodes table);
* The node you are starting will not write itself to the `nodes` table;
* The `/cluster` endpoint on each of your alive Kong nodes will give different results;
* Each node you start will believe itself to be in its own cluster.

I can work around this by using a short TTL such as 1 minute, and waiting 1 minute before trying to replace a dead Kong node. Another possible workaround is to put `kong stop` in an rc.shutdown script for when your *nix machine shuts down.

Here are repro steps using Docker. It uses the default 1 hour TTL and I'm using postgres 9.5 because that's what I use in AWS RDS (because of its superior default config compared to 9.4).

```
docker run -d --name kong-database -p 5432:5432 -e "POSTGRES_USER=kong" -e "POSTGRES_DB=kong" postgres:9.5
docker run -d --name kong -e "DATABASE=postgres" --link kong-database:kong-database -p 8000:8000 -p 8443:8443 -p 8001:8001 -p 7946:7946 -p 7946:7946/udp --security-opt seccomp:unconfined mashape/kong
# allow enough time for kong to start (check for a response from http://docker:8001/cluster)
docker stop kong
docker rm kong
```

Now we are going to set the IP address to some garbage value that does not exist. This is to simulate a Kong node being terminated without `kong stop`, and a new kong node starting with a different IP.

```
docker exec -it kong-database psql -U kong
kong=# update nodes set cluster_listening_address='1.2.3.4:7946';
```

Now start Kong again and get the output:

```
docker run -it --name kong -e "DATABASE=postgres" --link kong-database:kong-database -p 8000:8000 -p 8443:8443 -p 8001:8001 -p 7946:7946 -p 7946:7946/udp --security-opt seccomp:unconfined mashape/kong

[INFO] kong 0.8.3
[INFO] Using configuration: /etc/kong/kong.yml
[INFO] Setting working directory to /usr/local/kong
[INFO] database...........postgres host=kong-database database=kong user=kong port=5432
[INFO] dnsmasq............address=127.0.0.1:8053 dnsmasq=true port=8053
[INFO] serf ..............-profile=wan -rpc-addr=127.0.0.1:7373 -event-handler=member-join,member-leave,member-failed,member-update,member-reap,user:kong=/usr/local/kong/serf_event.sh -bind=0.0.0.0:7946 -node=6e6c12cd9931_0.0.0.0:7946_5dfdcec53bab494abe79b9fb822ec38f -log-level=err
[INFO] Trying to auto-join Kong nodes, please wait..
[WARN] Cannot join 1.2.3.4:7946. If the node does not exist anymore it will be automatically purged.
[WARN] Could not join the existing cluster
```

You will get a response from the cluster endpoint at http://docker:8001/cluster. But the `nodes` table has no mention of your new node; only the old dead entry:

```
docker exec -it kong-database psql -U kong
kong=# select * from nodes;
                            name                            | cluster_listening_address |     created_at
------------------------------------------------------------+---------------------------+---------------------
 062ae8f19597_0.0.0.0:7946_dd0b830a4582406a926649240952a1b7 | 1.2.3.4:7946              | 2016-06-06 10:25:11
(1 row)
```

Therefore, for every kong node you now start, they will each join their own cluster, and nodes will not be aware of each other.

Now we will try setting the TTL of the dead node to some time obviously in the past:

```
docker stop kong
docker rm kong
docker exec -it kong-database psql -U kong
kong=# update ttls set expire_at='2016-01-01 00:00:00' where primary_key_value='062ae8f19597_0.0.0.0:7946_dd0b830a4582406a926649240952a1b7';
kong=# \q
docker run -it --name kong -e "DATABASE=postgres" --link kong-database:kong-database -p 8000:8000 -p 8443:8443 -p 8001:8001 -p 7946:7946 -p 7946:7946/udp --security-opt seccomp:unconfined mashape/kong
```

Kong will now start successfully. New kong nodes will properly join the cluster, and purging will take place as per normal behaviour.

This scenario can arise in real life if you use AWS EC2 and a node in your auto scaling group is terminated. If your kong.yml TTL is longer than the time it takes EC2 to spin up a new node (which is a few minutes), you hit this bug.

While using a short TTL avoids the problem, I think the bug is still valid.

**Comment by dag24** [comment]
TL;DR if a node can't join the cluster, that node won't write itself to `nodes`.

**Comment by asknaskan** [comment]
I just looked into the code and logging the postgresql statements and found out, that the function `_auto_join()` from "kong/kong/cli/services/serf.lua#L109" contains a part, that deletes an node from database.
This should only happen at the start of kong, but it be happen an second time, if kong is not clustered to another node and before tries to join one cluster in "kong/kong/core/cluster.lua:#L46" cause condition #members < 2 will be true if kong is not clustered at this point.
That means, that kong tries to join a cluster, fails, adding itself on start to database, tries to reconnect to node and deleting itself again from database. In this case, the node has no possibility to create an new entry in the table again.

For fixing that, I easily just move the part of "kong/kong/cli/services/serf.lua#L109" to "kong/kong/cli/services/serf.lua#L214".

**Comment by thefosk** [comment]
@asknaskan could you submit a PR?

**Comment by asknaskan** [comment]
@thefosk Sure, here you are.

**Comment by thibaultcha** [comment]
Considering this resolved since the PR is merged.
